pkg/owner/owner.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package owner contains code for walking up the ownership chain,
+// starting with an arbitrary object. RBAC rules must allow GET access
+// to each object on the chain, at least including the starting
+// object, more when walking up more than one level.
+package owner
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Lookup walks up the ownership chain zero or more levels and returns an OwnerReference for the
+// object. The object identified by name, namespace and type is the starting point and is
+// returned when levels is zero. Only APIVersion, Kind, Name, and UID will be set.
+// IsController is always true.
+func Lookup(config *rest.Config, namespace, name string, gkv schema.GroupVersionKind, levels int) (*metav1.OwnerReference, error) {
+	c, err := client.New(config, client.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("build client: %v", err)
+	}
+
+	return lookupRecursive(c, namespace, name, gkv, levels)
+}
+
+func lookupRecursive(c client.Client, namespace, name string, gkv schema.GroupVersionKind, levels int) (*metav1.OwnerReference, error) {
```

Comment:
Would it make sense to have `apiVersion` and `kind` as parameters instead of `GroupVersionKind`? The recursive loop parses `gvk` from an object via `ParseGroupVersion` only to assemble it back into `APIVersion` / `Kind` in the next iteration.

In reality the loop will be executed twice at max, so I don't have a strong opinion.

Reply:
Yes, that makes sense. I picked gkv for the exported API because it is strongly typed instead of three loosely type strings, but here the code is local and strings are better.

Reply:
Done.